http/middleware/idempotency/inmem.go

```diff
@@ -0,0 +1,78 @@
+package idempotency
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+type InMemStore struct {
+	mu        sync.RWMutex
+	inFlight  map[string][]byte // key -> request signature
+	responses map[string]*StoredResponse
+}
+
+// NewInMemStore initializes an in-memory store.
+func NewInMemStore() *InMemStore {
+	return &InMemStore{
+		inFlight:  make(map[string][]byte),
+		responses: make(map[string]*StoredResponse),
+	}
+}
+
+func (s *InMemStore) InsertInFlight(_ context.Context, key string, requestSignature []byte) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// If we already have a marker or a stored response, we should decide
+	if _, ok := s.inFlight[key]; ok {
+		return errors.New("already in-flight") //nolint:err113 // it's mostly for testing
+	}
+
+	if _, ok := s.responses[key]; ok {
+		return errors.New("already completed") //nolint:err113 // it's mostly for testing
+	}
+
+	s.inFlight[key] = requestSignature
+
+	return nil
+}
+
+func (s *InMemStore) MarkComplete(_ context.Context, key string, resp *StoredResponse) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// We assume the request was in-flight, so we remove that marker
+	// and store the final response.
+	delete(s.inFlight, key)
+	s.responses[key] = resp
+
+	return nil
+}
```

Comment:
In the mark complete, we need to store the req fingerprint.
So we can check the conflict between incoming request and cache request.